kms/tpmkms/tpmkms.go

```diff
@@ -73,21 +74,34 @@ type TPMKMS struct {
 }
 
 type algorithmAttributes struct {
-	Type  string
-	Curve int
+	Type     string
+	Curve    int
+	Requires []tpm2.Algorithm
 }
 
 var signatureAlgorithmMapping = map[apiv1.SignatureAlgorithm]algorithmAttributes{
-	apiv1.UnspecifiedSignAlgorithm: {"RSA", -1},
-	apiv1.SHA256WithRSA:            {"RSA", -1},
-	apiv1.SHA384WithRSA:            {"RSA", -1},
-	apiv1.SHA512WithRSA:            {"RSA", -1},
-	apiv1.SHA256WithRSAPSS:         {"RSA", -1},
-	apiv1.SHA384WithRSAPSS:         {"RSA", -1},
-	apiv1.SHA512WithRSAPSS:         {"RSA", -1},
-	apiv1.ECDSAWithSHA256:          {"ECDSA", 256},
-	apiv1.ECDSAWithSHA384:          {"ECDSA", 384},
-	apiv1.ECDSAWithSHA512:          {"ECDSA", 521},
+	apiv1.UnspecifiedSignAlgorithm: {"RSA", -1, []tpm2.Algorithm{tpm2.AlgRSA}},
+	apiv1.SHA256WithRSA:            {"RSA", -1, []tpm2.Algorithm{tpm2.AlgRSA, tpm2.AlgSHA256}},
+	apiv1.SHA384WithRSA:            {"RSA", -1, []tpm2.Algorithm{tpm2.AlgRSA, tpm2.AlgSHA384}},
+	apiv1.SHA512WithRSA:            {"RSA", -1, []tpm2.Algorithm{tpm2.AlgRSA, tpm2.AlgSHA512}},
+	apiv1.SHA256WithRSAPSS:         {"RSA", -1, []tpm2.Algorithm{tpm2.AlgRSAPSS, tpm2.AlgSHA256}},
+	apiv1.SHA384WithRSAPSS:         {"RSA", -1, []tpm2.Algorithm{tpm2.AlgRSAPSS, tpm2.AlgSHA384}},
+	apiv1.SHA512WithRSAPSS:         {"RSA", -1, []tpm2.Algorithm{tpm2.AlgRSAPSS, tpm2.AlgSHA512}},
+	apiv1.ECDSAWithSHA256:          {"ECDSA", 256, []tpm2.Algorithm{tpm2.AlgECDSA, tpm2.AlgSHA256}},
+	apiv1.ECDSAWithSHA384:          {"ECDSA", 384, []tpm2.Algorithm{tpm2.AlgECDSA, tpm2.AlgSHA384}},
+	apiv1.ECDSAWithSHA512:          {"ECDSA", 521, []tpm2.Algorithm{tpm2.AlgECDSA, tpm2.AlgSHA512}},
+}
+
+var preferredAlgorithm = []apiv1.SignatureAlgorithm{
+	apiv1.ECDSAWithSHA384,
+	apiv1.ECDSAWithSHA512,
+	apiv1.ECDSAWithSHA256,
+	apiv1.SHA512WithRSA,
+	apiv1.SHA384WithRSA,
+	apiv1.SHA256WithRSA,
+	apiv1.SHA512WithRSAPSS,
+	apiv1.SHA384WithRSAPSS,
+	apiv1.SHA256WithRSAPSS,
```

Comment:
I considered just having the keys of `signatureAlgorithmMapping` be the preferred order, but that seemed a little too terse. Happy to take another approach here.

Reply:
I think the functionality is nice, but the order defined here might be too opinionated and possibly artificial to impose here, especially when it comes to the order of hashing algorithms. Given the hash calculations are performed by the TPM, higher numbers are likely to have an impact on performance that is more noticeable than when performed on a CPU.

Reply:
d883afe